onnxruntime/core/optimizer/pre_shape_node_elimination.cc

```diff
@@ -0,0 +1,55 @@
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/pre_shape_node_elimination.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status PreShapeNodeElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  // Remove Output Edges.
+  graph_utils::RemoveNodeOutputEdges(graph, node);
+  const Node* p_input_node = graph_utils::GetInputNode(node, 0);
+
+  // Get mutable input node
+  Node& input_node = *graph.GetNode(p_input_node->Index());
+  const int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(input_node, node.MutableInputDefs()[0]->Name());
+
+  // Update InputDefs of the next shape nodes.
+  auto output_nodes = graph.GetMutableConsumerNodes(node.OutputDefs()[0]->Name());
+  for (Node* next_node : output_nodes) {
+    for (auto& input_def : next_node->MutableInputDefs()) {
+      input_def = input_node.MutableOutputDefs()[output_idx];
+    }
+  }
+
+  graph.RemoveNode(node.Index());
+  rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
+
+  return Status::OK();
+}
+
+bool PreShapeNodeElimination::SatisfyCondition(const Graph& graph, const Node& node, const logging::Logger& logger) const {
+  if (!graph_utils::CanRemoveNode(graph, node, logger)) {
+    return false;
+  }
+
+  auto output_nodes = graph.GetConsumerNodes(node.OutputDefs()[0]->Name());
```

Comment:
I'm looping over output nodes and not edges.
can you elaborate on which part you want this to be used?

Reply:
It's similar

for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it) {

Reply:
NodeConstIterator is used for const access to Node instances, and since we need to modify the output nodes, we cannot use that iterator directly.

Reply:
we can. you take the const node, get the node index, then graph.GetNode() to get the writable instance. 

 Node* GetNode(NodeIndex node_index)

Reply:
Once we delete edges using   graph_utils::RemoveNodeOutputEdges(graph, node);
we lose access to node.OutputNodesBegin() and node.OutputNodesEnd()

we also need to remove edges before changing the inputdefs, otherwise if we change inputdefs before, we would get an error of mismatch types when removing edge.